File: balboa-agent/src/main/scala/com/socrata/balboa/agent/MetricsRecord.scala
package com.socrata.balboa.agent

import java.lang.{Double => JavaDouble, Long => JavaLong}
import java.util.regex.Pattern

import com.socrata.balboa.metrics.Metric
import com.socrata.balboa.metrics.Metric.RecordType
import scodec.bits.{BitVector, ByteVector}
import scodec.codecs._
import scodec.{Codec, _}
/**
  * A Metrics Record is an immutable class that represents a metric
  * that occurred with a specific entity-id, name, value, type, and time.
  *
  * Created by michaelhotan on 2/2/16.
  * Converted from Java to Scala using scodec for decoding on 1/9/2017
  */
case class MetricsRecord(timestamp: Long, entityId: String, name: String, value: Number, metricType: RecordType)

/**
  * MetricsRecord Codec
  * <ul>
  * <li>0xff - single byte - Beginning mark of a single metrics entry</li>
  * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of timestamp of type long.</li>
  * <li>0xfe - single byte - end of timestamp sequence</li>
  * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the entity id</li>
  * <li>0xfe - single byte - end of entity id byte sequence</li>
  * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric name</li>
  * <li>0xfe - single byte - end of metric name byte sequence</li>
  * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric value.
  * The metric value is of type Number</li>
  * <li>0xfe - single byte - end of metric value byte sequence</li>
  * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric type.
  * See [[com.socrata.balboa.metrics.Metric.RecordType]]</li>
  * <li>0xfe - single byte - end of metric type byte sequence</li>
  * </ul>
  * </b>
  */

object MetricsRecord {
  private val integerPattern: Pattern = "-?[0-9]+".r.pattern

  private val startByte = 0xff.toByte
  private val separatorByte = 0xfe.toByte
  private def byteTerminatedCodec[A](separatorByte: Byte, codec: Codec[A]) =
    filtered(codec, new Codec[BitVector] {
      private val separator = BitVector(separatorByte)

      override def sizeBound: SizeBound = SizeBound.unknown

      override def encode(bits: BitVector): Attempt[BitVector] = Attempt.successful(bits ++ separator)

      override def decode(bits: BitVector): Attempt[DecodeResult[BitVector]] = {
        bits.bytes.indexOfSlice(separator.bytes) match {
          case -1 => Attempt.failure(Err("Does not contain a '0x%02x' separator byte.".format(separatorByte)))
          case idx: Long => Attempt.successful(DecodeResult(bits.take(idx * 8L), bits.drop(idx * 8L + 8L)))
        }
      }
    })

  private val metricFieldStringCodec = byteTerminatedCodec(separatorByte, utf8)

  private val valueCodec: Codec[Number] = metricFieldStringCodec.exmap[Number]({ rawValue: String =>
    if (rawValue.equalsIgnoreCase("null")) {
      Attempt.successful(null) // scalastyle:ignore
    } else {
      try {
        if (integerPattern.matcher(rawValue).matches) {
          Attempt.successful(JavaLong.valueOf(rawValue))
        } else {
          Attempt.successful(JavaDouble.valueOf(rawValue))
        }
      }
      catch {
        case _: NumberFormatException =>
          Attempt.failure(Err(s"Unable to decode raw value $rawValue into Number format"))
      }
    }
  }, n => Attempt.successful(n.toString))

  private val timestampCodec: Codec[Long] = metricFieldStringCodec.xmap[Long](_.toLong, _.toString)

  private val metricTypeCodec = metricFieldStringCodec
    .xmap[RecordType](s => Metric.RecordType.valueOf(s.toUpperCase), _.toString)

  implicit val codec: Codec[MetricsRecord] = {
    ("metric start indicator" | byteTerminatedCodec(startByte, bytes).unit(ByteVector.empty)) :~>:
      ("timestamp" | timestampCodec) ::
      ("entityId" | metricFieldStringCodec) ::
      ("name" | metricFieldStringCodec) ::
      ("value" | valueCodec) ::
      ("metricType" | metricTypeCodec)
  }.as[MetricsRecord]
}

File: balboa-agent/src/main/java/com/socrata/balboa/agent/MetricsRecord.java
package com.socrata.balboa.agent;

import com.socrata.balboa.metrics.Metric;

/**
 * A Metrics Record is an immutable class that represents a metric
 * that occurred with a specific entity-id, name, value, type, and time.
 *
 * Created by michaelhotan on 2/2/16.
 */
public class MetricsRecord {

    private final String entityId;
    private final String name;
    private final Number value;
    private final Metric.RecordType type;
    private final long timestamp;

    public MetricsRecord(String entityId,
                         String name,
                         Number value,
                         long timestamp,
                         Metric.RecordType type)
    {
        this.entityId = entityId;
        this.name = name;
        this.value = value;
        this.timestamp = timestamp;
        this.type = type;
    }

    public String getEntityId() {
        return entityId;
    }

    public String getName() {
        return name;
    }

    public Number getValue() {
        return value;
    }

    public Metric.RecordType getType() {
        return type;
    }

    public long getTimestamp() {
        return timestamp;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MetricsRecord that = (MetricsRecord) o;

        return timestampsEqual(that)
                && entityIdsEqual(that)
                && namesEqual(that)
                && valuesEqual(that)
                && type == that.type;
    }

    @Override
    public int hashCode() {
        int result = entityId != null ? entityId.hashCode() : 0;
        result = 31 * result + (name != null ? name.hashCode() : 0);
        result = 31 * result + (value != null ? value.hashCode() : 0);
        result = 31 * result + (type != null ? type.hashCode() : 0);
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        return result;
    }

    @Override
    public String toString() {
        return "MetricsRecord{" +
                "entityId='" + entityId + '\'' +
                ", name='" + name + '\'' +
                ", value=" + value +
                ", type=" + type +
                ", timestamp=" + timestamp +
                '}';
    }

    private boolean timestampsEqual(MetricsRecord other) {
        return (timestamp == other.timestamp);
    }

    private boolean entityIdsEqual(MetricsRecord other) {
        if (entityId == null) {
            return other.entityId == null;
        } else {
            return entityId.equals(other.entityId);
        }
    }

    private boolean namesEqual(MetricsRecord other) {
        if (name == null) {
            return other.name == null;
        } else {
            return name.equals(other.name);
        }
    }

    private boolean valuesEqual(MetricsRecord other) {
        if (value == null) {
            return other.value == null;
        } else {
            return value.equals(other.value);
        }
    }
}
